### ISSUE TITLE
Only f32/i32/i64 execute: f16, bf16, i8, i4 are declaration-only — every dtype vLLM serves Llama in

### ISSUE THREAD

**Issue by hiraditya** [vx-lang/Vx#320]
`f32`, `i32` and `i64` are the only numeric element types that survive codegen. **Every dtype an LLM is actually served in — `f16`, `bf16`, `fp8`, `i8`, `i4` — can be declared and capacity-checked but cannot be executed.**

Found while scoping the GPU demo (#319). It bounds every dtype claim in the MLSys fleet work to *admission* rather than *execution*, and it is on the critical path for a Llama/FlashAttention demo whose numbers anyone would want to show.

## Measured

Identical program per dtype — declare a `Tensor<T>([4,4])`, write elements, read one back:

```rust
fn main() -> i32 {
  let mut a = Tensor<T>([4, 4]);
  for i in 0..4 { for j in 0..4 { a[i][j] = <lit>; } }
  print(a[0][0]);
  return 0;
}
```

| dtype | What vLLM uses it for on Llama / FA | Declare + admit | Execute |
|---|---|---|---|
| `f32` | reference / baseline only | ✅ | ✅ |
| `i32`, `i64` | token ids, indices, logits argmax | ✅ | ✅ |
| **`f16`** | **the default Llama serving dtype** | ✅ | ❌ `Unsupported MLIR element type in: f16` |
| **`bf16`** | **the other default Llama serving dtype** | ✅ | ❌ `Unsupported MLIR element type in: bf16` |
| `f8e4m3` | FP8 weights/activations (W8A8) and KV cache on H100/B200 | ✅ | ❌ `fp8 element types are capacity/declaration-only; fp8 codegen is tracked in hiraditya/Vx.1#249` |
| `f8e5m2` | FP8 KV cache | ✅ | ❌ same as above |
| `i8`, `u8` | INT8 quantisation (W8A8) | ✅ | ❌ `Unsupported MLIR element type in: i8` |
| **`i4`, `u4`** | **AWQ / GPTQ 4-bit — the most common way Llama is actually deployed** | ✅ | ❌ `Unsupported MLIR element type in: i4` |
| `i16`, `u16` | rare in this setting | ✅ | ❌ `Unsupported MLIR element type in: i16` |

`u8`/`u4` report through their signed spelling (`i8`/`i4`), so the unsigned variants are not separately handled either.

## Why declaration works and execution does not

`fleet/admit.vx` emits `memref<?x?xf16>` and admits correctly, which is what makes this easy to miss: the **type** reaches MLIR, so capacity arithmetic over `f16` is real and correct. It is *element access* that fails. `admit.vx` only declares and allocates — it never reads or writes an element — so the whole fleet-admission matrix is unaffected and its results stand.

There are two distinct mechanisms in [`extract_mlir_element_type`](https://github.com/hiraditya/Vx/blob/main/src/codegen/lower/mod.rs#L134):

1. **`f16` / `bf16` are already in the accept list** (lines 155-156) and still fail. They therefore cannot be reaching the `match` at all: the function early-returns at **line 141** for any string that is not wrapped in `memref<…>` / `tensor<…>`, and the error text shows a *bare* `f16` arriving. A scalar element type extracted from a load is being handed to a function that only accepts container spellings.
1. **`i8`, `i16`, `i4`, `u8`, `u4` are absent from the `match` entirely** (line 154-163), so they fail at line 162 regardless of how they arrive.

Both need fixing; (1) is likely a few lines, since the accept list already names the types.

The fp8 case is **not** a bug — it is deliberately gated with a message naming hiraditya/Vx.1#249. That is the right behaviour (fail closed, point at the tracker) and only listed here for completeness of the dtype picture.

## Why it matters beyond tidiness

- **2× on a bandwidth-bound workload.** Batch-1 Llama decode is memory-bandwidth-bound; the GEMMs degenerate to GEMV. Running `f32` where vLLM runs `f16`/`bf16` reads twice the bytes per token, which is a straight 2× penalty *before any kernel is involved*. On a cuBLAS-backed demo this would be the single largest term in any performance delta, and it has nothing to do with kernel quality.
- **It rules out the deployment-realistic configurations.** AWQ/GPTQ 4-bit is how Llama is most commonly served. A fleet-admission story that can admit an `i4` configuration but never execute one has a visible gap between what it reasons about and what it can run.
- **cuBLAS is ready for these and we are not.** `HGEMM` / `cublasGemmEx` cover fp16/bf16/int8 natively, so the blocker is entirely on the Vx side.

## Acceptance

- The program above executes and prints the correct value for `f16`, `bf16`, `i8`, `i16`, `i4` (and the `u*` spellings).
- A corpus fixture per dtype, asserting the *value*, not just that MLIR was emitted — the existing Llama fixtures only check `emit-mlir` produces a module, which is how this stayed invisible.
- `fp8` continues to fail closed with its hiraditya/Vx.1#249 message rather than silently producing wrong results.
- The distinction stays documented: capacity admission over a dtype does not imply execution in it.

## Related

- hiraditya/Vx#319 — GPU demo; this is a Track-A blocker.
- hiraditya/Vx.1#249 — FP8/FP6/FP4 codegen + block scaling.
- hiraditya/Vx.1#280 — added the fp8 element types (declaration side, landed).
- hiraditya/Vx.1#290 — the MLSys readiness table lists `f16`/`bf16` and `i4` packing under "already there"; that is accurate for **capacity math** and should be read as such.

**Comment by hiraditya** [comment]
## Measured: f16 executes today. The gap was printing, not compute.

Fixed in aa8f1c6, but the more useful outcome is a correction to this issue's premise, because it changes M1's ordering in hiraditya/Vx.1#321.

This issue reports `f16`/`bf16`/`i8` as "Declare + admit ✅ / Execute ❌", and the follow-up comment puts the dtype work **ahead of cuBLAS wiring** on the critical path, reasoning that serving in `f32` is "a straight 2× penalty from reading twice the bytes."

Probed each claim directly on the current tree (Linux, LLVM/MLIR 22):

| Probe | Result |
|---|---|
| `f16` scalar arithmetic — `1.5 + 2.5` | **4** ✅ |
| `f16` tensor: `Tensor<f16>([2,2])`, store, load | **1.5** ✅ |
| `f16` matmul: `a @ b`, 2×2 through `linalg` | **4** ✅ (`lhs=memref<?x?xf16>, rhs=memref<?x?xf16>`) |
| `i8` scalar arithmetic — `3 + 4` | **7** ✅ |
| `print(x)` where `x : f16` | ❌ `Unsupported MLIR element type in: f16` |

So the compute path — which is what a serving program's bandwidth bill is made of — already runs in f16, including the GEMM. The only failure was `print`, a diagnostic convenience.

**The 2× bandwidth penalty this issue prices in was never being paid.** A demo built on f16 compute was available all along; what would have failed is printing a scalar during debugging.

### The actual mechanism

The diagnosis in the issue body is right about *where* but not about *what*. `extract_mlir_element_type` (`src/codegen/lower/mod.rs`) does return early unless the string is `memref<...>` or `tensor<...>` — but that function is only reached on the **memref print path**. The scalar fast-path above it handles `i32`/`i64`/`f32`/`f64` and nothing else, so a bare `f16` fell past it into memref handling, which then correctly reported that `f16` is not a memref.

Fix: widen before printing — `arith.extf` to `f32` for `f16`/`bf16`, `arith.extsi` to `i32` for `i8`/`i16` — and use the helper that already exists. Adding `f16` to the accept list of `extract_mlir_element_type` would not have helped, since `printMemrefF16` does not exist either.

Verified: `f16` 4.5, `bf16` 2.5, `i8` 7, `i16` 300 all print. Backend suites and the 106 flat-vs-AST differential tests pass.

### What is still genuinely blocked

- **fp8** — `src/hir/flatten.rs` returns `None` for `F8E4M3`/`F8E5M2`. The issue is right that this gate is deliberate.
- **i4/u4** — no MLIR scalar mapping, which matters for the AWQ/GPTQ deployments this issue names as the most common.
- Neither is on the path to an f16 demo.

### Consequence for hiraditya/Vx.1#321

M1's ordering can revert to what hiraditya/Vx#319 originally had: **device residency → cuBLAS dispatch → program**, with dtype work off the critical path. Residency is the performance story, as hiraditya/Vx#319 said before this issue displaced it.

Worth stating plainly, since this is the second time in this campaign that a documented blocker dissolved on measurement (the first: `llama2_v2.vx` being the recommended serving base, when in fact it prints pointers — hiraditya/Vx.1#323). Both were reported from reading rather than running. The cheap correction is to run the three-line probe before pricing the work.
